### ISSUE TITLE
File close when save

### ISSUE THREAD

**Issue by francescofinotto** [mono/monodevelop#6488]
when i press ctrl + s for saving changes in a file, the source code editor of that file close.

> VS bug [#718767](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/718767)

**Comment by Therzok** [comment]
Related to FSW I guess. Don't see it on mac, so it's possibly a linux/windows thing

**Comment by francescofinotto** [comment]
i'm using monodevelop on windows 10. just rebuild

**Comment by mickeyjf** [comment]
Also on Widows 7. I hoped it was just a setting I did not know about.

**Comment by Therzok** [comment]
I'll take a look soon, it's bugging me too when I have to work on windows. 👍

**Comment by piratesephiroth** [comment]
This is VERY annoying. Happens all the time and it also closes the file instead of running when we hit F5

**Comment by Therzok** [comment]
Moved on current sprint.

**Comment by Therzok** [comment]
Initial investigation here is that we seem to redirect a document to contain the temporary save file thus when saving it, it gets closed.

**Comment by Therzok** [comment]
Seems like the OS FileSystemWatcher is telling us that `Document.cs` is renamed to `Document.cs~<hash>.tmp` so we keep track of the tmp file then, and when the save is done, it's deleted, so we close the document.

**Comment by Therzok** [comment]
I opened an Initial PR to fix this: https://github.com/mono/monodevelop/pull/6850/files

But I will discard that work and start a different approach, because, as stated in the PR above, we're going to replace that code we have with a library, so we'll end up having the same problem in the future.

Next up, I'm going to tackle this problem and solve it step by step, as it is not something I'd like to change without testing the waters first.

* [x] Step 1: Optimized and cleaned up a bit of the code so it's easier to change - https://github.com/mono/monodevelop/pull/6864
* [ ] Step 2: Fix EventChainArgs merging for FileCopyEventInfo when IsExternal differs between the two
* [ ] Step 3: Try to get rid of all the generics and indirection there by having just one base `FileEventInfo` class, allowing us to do easier processing of data - i.e. get rid of EventData's hard to follow abstraction
* [ ] Step 4: Introduce an `EventQueue.Processor` class that simplifies events, handles deduplication and merges the EventArgChains. See [0] for some examples. 
* [ ] Step 5: Properly detect the `File.Replace` [1] pattern and trigger events that would not allow the bug to happen.

[0] - Patterns to reduce:

Create A, Delete A -> NOP
Change A, Change A -> Change A
Delete A, Create A -> Change A

[1] - File.Replace(A, B) pattern:
Copy B -> B.tmp
Move A -> B
Remove tmp

Becomes move A -> B

**Comment by Therzok** [comment]
Opened a paper doc with the proposals and steps for the changes:
https://paper.dropbox.com/doc/FileService-API-breakages-and-optimizations-Lmx4sBpGbT2V0vd2IGD1S

**Comment by xamarin-release-manager** [comment]
> [sync] [VS-28] Comment by **xamarinc**



Fixed in version 8.0.0.1557 (master)



Author: Marius Ungureanu  
 Commit: 25d91fc2ad8756f270653bbfd03023d096436580 (mono/monodevelop)
